File: Trees/LC_235.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.29

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return search(root, p->val, q->val);
    }
private:
    TreeNode* search(TreeNode* root, int pval, int qval){
        if (root == nullptr){
            return root;
        }
        if (root->val < pval && root->val < qval){
            return search(root->right, pval, qval);
        }
        else if(root->val > pval && root->val > qval){
            return search(root->left, pval, qval);
        }
        else{
            return root;
        }
    }
};

File: Trees/LC_235.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.29

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def check(root, p, q):
            if not root:
                return None
            if root.val < p.val and root.val < q.val:
                return check(root.right, p, q)
            elif root.val > p.val and root.val > q.val:
                return check(root.left, p, q)
            else:
                return root
        return check(root, p, q)